### PR TITLE
Improved the extendability of the SchemaView and DataGridView.

### PR DESCRIPTION
**Pull request by asheshv**
Restructured the SchemaView & DataGridView modules for ease of maintenance and applied the single responsibility principle (wherever is applicable).

**SchemaView**
- Split the code based on the functionality, and responsibility.
- Introduced a new View 'InlineView' instead of using the 'nextInline' configuration of the fields to have a better, and manageable code.
- Using the separate class 'SchemaState' for managing the data and states of the SchemaView (separated from the 'useSchemaState' custom hook).
- Introduced three new custom hooks 'useFieldValue', 'useFieldOptions', 'useFieldError' for the individual control to use for each Schema Field.
- Don't pass value as the parameter props, and let the 'useFieldValue' and other custom hooks to decide, wheather to rerender the control itself or the whole dialog/view. (single responsibilty principle)
- Introduced a new data store with subscription facility.
- Moving the field metadata (option) evaluation to separate place for better management, and each option can be defined for a particular kind of field (for example - colleciton, row, cell, general, etc).
- Allow to provide custom control for all kind of Schema field.

**DataGridView**
- Same as SchemaView, split the DataGridView call in smaller and manageable chunks. (For example - grid, row, mappedCell, etc).
- Use context based approach for providing the row and table data instead of passing them as parameter to each and every component separately.
- Have a facility to extend this feature separately in future. (for example - selecatable cell, column grouping, etc.)
- Separated the features like deletable, editable, reorder, expandable etc. cells using the above feature support.
- Added ability to provide the CustomHeader, and CustomRow through the Schema field, which will extend the ability to customisation better.
- Removed the 'DataGridViewWithHeaderForm' as it has been achieved through providing 'CustomHeader', and also introduced 'DataGridFormHeader' (a custom header) to achieve the same feature as 'DataGridViewWithHeaderForm'.

Also - Modified the test-cases to use separate 'Schema' objects during testing the rendering of create, edit, and properties views as BaseUISchema instance share the same SchemaState object.

### CODE REVIEW

**Comment by asheshv** [comment]
Will work on the feature test failures.

**Review by adityatoshniwal — COMMENTED** [review]
1. Open an existing table and change column to PK. It crashes.
2. Partitioned table properties tab is crashing, even on v8.11.
3. Auto focus first input is not working.
4. Go to security tab and add a row. Then select a grantee and click on privileges. Notice - it requires 2 clicks to open
5. 
image

File: web/pgadmin/browser/server_groups/servers/databases/schemas/types/static/js/type.ui.js [hunk 1]
```diff
@@ -1024,7 +1024,7 @@ class DataTypeSchema extends BaseUISchema {
     return [{
       id: 'type',
       label: gettext('Data Type'),
-      group: gettext('Definition'),
+      group: gettext('Data Type'),
```
[the remark this answers is not in the document]
Comment (asheshv) [reply web/pgadmin/browser/server_groups/servers/databases/schemas/types/static/js/type.ui.js:1027]: It's unintentional.
I will revert it back.

File: web/pgadmin/browser/server_groups/servers/databases/schemas/tables/static/js/table.ui.js [hunk 2]
```diff
@@ -274,46 +276,47 @@ export class LikeSchema extends BaseUISchema {
         id: 'like_default_value', label: gettext('With default values?'),
         type: 'switch', mode: ['create'], deps: ['like_relation'],
         disabled: this.isRelationDisable, depChange: (...args)=>obj.resetVals(...args),
-        inlineNext: true,
+        inlineGroup: 'like_relelation',
```
[the remark this answers is not in the document]
Comment (asheshv) [reply web/pgadmin/browser/server_groups/servers/databases/schemas/tables/static/js/table.ui.js:279]: Agree.
I will change.

File: web/pgadmin/static/js/SchemaView/SchemaState/SchemaState.js [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+      })
+    ) state.setError({});
+
+    state.data = sessData;
+    state.changes = state.Changes();
+    state.updateOptions();
+    state.onDataChange && state.onDataChange(state.isDirty, state.changes);
+  }
+
+  Changes(includeSkipChange=false) {
```
[the remark this answers is not in the document]
Comment (asheshv) [reply web/pgadmin/static/js/SchemaView/SchemaState/SchemaState.js:228]: There is already a variable changes, hence - used 'C' capital here.
But - I agree - we should not break the coding style.
I will look for alternative implementation.

File: web/pgadmin/static/js/SchemaView/DataGridView/features/fixedrows.jsx [hunk 4]
```diff
@@ -0,0 +1,46 @@
+/////////////////////////////////////////////////////////////
```
[the remark this answers is not in the document]
Comment (asheshv) [reply web/pgadmin/static/js/SchemaView/DataGridView/features/fixedrows.jsx:1]: Sure.

**Comment by asheshv** [comment]
> 1. Open an existing table and change column to PK. It crashes.

We don't allow to change the PK for a column. Can you please share exact steps?

> 2. Partitioned table properties tab is crashing, even on v8.11.

Fixed.

> 3. Auto focus first input is not working.

Yet to work on it.

> 4. Go to security tab and add a row. Then select a grantee and click on privileges. Notice - it requires 2 clicks to open

Not reproducible.

> 
> image

Yet to work on it.

**Comment by adityatoshniwal** [comment]
Hi @asheshv,
Attaching screen recordings:

Trying to add column in existing table/or switch a column to PK where there is no PK.
https://github.com/user-attachments/assets/2efcd99f-be1a-43a5-9b2c-ff957afaac0d

In this, when grantee is changed, the 2 clicks are required to open privileges.
https://github.com/user-attachments/assets/fdcfee63-dfed-4979-b644-dc00150b08ab

**Comment by asheshv** [comment]
> Hi @asheshv, Attaching screen recordings:
> 
> Trying to add column in existing table/or switch a column to PK where there is no PK. https://github.com/user-attachments/assets/2efcd99f-be1a-43a5-9b2c-ff957afaac0d
So - it was failing while creating a table, and not existing table. :)
Fixed.

> In this, when grantee is changed, the 2 clicks are required to open privileges. https://github.com/user-attachments/assets/fdcfee63-dfed-4979-b644-dc00150b08ab

**Comment by adityatoshniwal** [comment]
> So - it was failing while creating a table, and not existing table. :) Fixed.

No, I'm adding column to existing table.

**Comment by asheshv** [comment]
> > In this, when grantee is changed, the 2 clicks are required to open privileges. https://github.com/user-attachments/assets/fdcfee63-dfed-4979-b644-dc00150b08ab
> 
> Looking into it.

Done!

**Comment by asheshv** [comment]
@adityatoshniwal All issues are done. Please check now.

**Comment by adityatoshniwal** [comment]
> @adityatoshniwal All issues are done. Please check now.
> 
> EDIT: There is one issue remaining - first element selection is not covered.

I missed in the review last time - SchemaStateContext, SCHEMA_STATE_ACTIONS variables are duplicated in 2 files. Please remove that as well.

**Comment by asheshv** [comment]
> > @adityatoshniwal All issues are done. Please check now.
> > EDIT: There is one issue remaining - first element selection is not covered.
> 
> I missed in the review last time - SchemaStateContext, SCHEMA_STATE_ACTIONS variables are duplicated in 2 files. Please remove that as well.

As discussed, this has already been taken care in this PR.

**Comment by asheshv** [comment]
> There is one issue remaining - first element selection is not covered.

Done.

**Pull request merged**